### PR TITLE
[AIRFLOW-3211] Reattach to GCP Dataproc jobs upon Airflow restart 

### PR DESCRIPTION
**Pull request by jj-ian**
This change allows Airflow to reattach to existing Dataproc jobs upon scheduler restart. Previously, if the Airflow scheduler restarts while it's running a job on GCP Dataproc, it'll lose track of that job, mark the task as failed, and eventually retry. However, the jobs may still be running on Dataproc and maybe even finish successfully. So when Airflow retries and reruns the job, the same job will run twice. This can result in issues like delayed workflows, increased costs, and duplicate data.

The fix also helps in cases where the task fails for some other reason: for example, sometimes the Dataproc service is unavailable, and the Airflow task will fail because it can't check the status of the job, but the job is actually still running correctly.

We've been using this solution successfully at Etsy for the past 6 months. 

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. 
  - https://issues.apache.org/jira/browse/AIRFLOW-3211

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

My change has Airflow query the Dataproc API before submitting a job to see if the job is already running on the cluster. If a job with a matching task ID is already running on the cluster AND is in a recoverable state (like RUNNING or DONE), then Airflow will reattach itself to the existing job on Dataproc instead of resubmitting a new job to the cluster. If the job on the cluster is in an irrecoverable state like ERROR, Airflow will resubmit the job.

To see this change in action:

Setup:
1. Set up a GCP Project with the Dataproc API enabled
2. Install Airflow.
3. In the box that's running Airflow, `pip install google-api-python-client oauth2client`
4. Start the Airflow webserver. In the Airflow UI, Go to Admin->Connections, edit the `google_cloud_default` connection, and fill in the Project Id field with your project ID.

To reproduce:

1. Install this DAG in the Airflow instance: https://github.com/GoogleCloudPlatform/python-docs-samples/blob/b80895ed88ba86fce223df27a48bf481007ca708/composer/workflows/quickstart.py Set up the Airflow variables as instructed at the top of the file.
2. Start the Airflow scheduler and webserver if they're not running already. Kick off a run of the above DAG through the Airflow UI. Wait for the cluster to spin up and the job to start running on Dataproc.
3. While the job's running, kill the scheduler. Wait 5 seconds or so, and then start it back up.
4. Airflow will retry the task and reattach to the existing task already on Dataproc. Look at the Airflow logs to observe "Reattaching to previously-started DataProc job run_dataproc_... (in state RUNNING)." Click on the cluster in Dataproc to observe that only the single job is running; a duplicate job has not been submitted.
5. Observe that, when the job finishes, Airflow detects the completion successfully and runs the downstream cluster delete operation.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added the following tests to `tests/contrib/hooks/test_gcp_dataproc_hook.py`:

When submitting a new job to Dataproc:
- If a job with the same task ID is already running on the cluster, don't resubmit the job.
- If the first matching job found on the cluster is in an irrecoverable state, keep looking for a job in a recoverable state to reattach to on the cluster. This ensures that Airflow will prioritize recoverable jobs when looking for jobs to reattach to on the cluster.
- If there are jobs running on the cluster, but none of them have the same task ID as the job we're about to submit, then submit the new job.
- If there are no other jobs already running on the cluster, then submit the job.
- If a job with the same task ID finished with error on the cluster, then resubmit the job for retry.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by kaxil** [comment]
cc @fenglu-g

File: airflow/contrib/hooks/gcp_dataproc_hook.py [hunk 1]
```diff
@@ -33,12 +33,82 @@ def __init__(self, dataproc_api, project_id, job, region='global',
...
+        except KeyError:
+            self.log.error('Job to submit is incorrectly configured.')
+            raise
+
+        jobs_on_cluster_response = dataproc_api.projects().regions().jobs().list(
+            projectId=self.project_id,
+            region=self.region,
+            clusterName=cluster_name).execute()
+
+        UUID_LENGTH = 9
```
Comment (ashb) [inline airflow/contrib/hooks/gcp_dataproc_hook.py:52]: 9 characters is not long enough for a UUID - is this just a random identifier?

**Comment by ashb** [comment]
This PR seems sensible - though I would have fixed this by changing your deploy process to not kill your workers - you don't need to do that to pick up the latest version of DAGs. 

(Airflow is very careful to not load DAG definitions in to any long running process, so each time a task form a dag runs it will pick up the latest definition from disk)

File: airflow/contrib/hooks/gcp_dataproc_hook.py (hunk 1 above)
Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: The Airflow Dataproc job builder truncates the UUID to 8 characters [here](https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/hooks/gcp_dataproc_hook.py#L94), so the task name submitted to Airflow consists of the task ID + underscore + an 8 digit ID. The 9 characters in my PR represents the 8 UUID characters plus the underscore.

**Comment by jj-ian** [comment]
Hi @ashb, thank you for the review. I see what you mean, but we wanted to ensure that Airflow keeps track of Dataproc jobs even if the machine that hosts Airflow restarts for reasons outside our control. Our Airflow is hosted on machines that restart regularly for system updates (specifically, [GCP App Engine Flex instances that are restarted weekly by Google to install OS updates](https://cloud.google.com/appengine/docs/flexible/python/how-instances-are-managed)).

**Comment by ashb** [comment]
Yeah, I can see this being useful. I don't know enough about GCP/DataProc in Airflow to review this further, sorry :)

**Comment by jj-ian** [comment]
@ashb No worries and thank you! Oh, and the fix also helps in cases where the task fails for some other reason: for example, sometimes the Dataproc service is unavailable, and the Airflow task will fail because it can't check the status of the job, but the job is actually still running correctly.

**Comment by jj-ian** [comment]
@fenglu-g Hi Feng, this is the fix I emailed you about on February 27, 2018 with the subject line "Airflow losing track of running Dataproc jobs upon Airflow restart". Looking forward to any feedback you may have on this. Thank you!

File: airflow/contrib/hooks/gcp_dataproc_hook.py (hunk 1 above)
Comment (fenglu-db) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: This seems to be a bit error prone. How about the following? Let's modify the submit job operator interface by taking a new parameter job_dedupe_regex, when set, the hook will match existing jobs with this job_dedupe_regex and skip a new submission if a new match is found and other conditions are met. 

This way, the hook implementation is entirely decoupled from the operator implementation re: job-name and open to other ways of deduping dataproc jobs.

**Comment by fenglu-db** [comment]
@jj-ian left one comment, PTAL. Thanks for the PR, this is going to be very useful.

File: airflow/contrib/hooks/gcp_dataproc_hook.py (hunk 1 above)
Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: Thank you @fenglu-g. Just to clarify, do you mean that I should modify the [`DataProcHook.submit`](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L229) and [`_DataProcJob.__init__`](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L31) functions to take in the new regex param like so?

```
class _DataProcJob(LoggingMixin):
    def __init__(other args..., job_dedupe_regex=None):

class DataProcHook(GoogleCloudBaseHook):
    def submit(other args..., job_dedupe_regex=[by default, regex that dedupes by matching the task ID]):
```
And then have `submit()` pass the regex along to the _DataProcJob init [here](https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L230
), because the deduping happens during the init:

```
def submit(other args..., job_dedupe_regex=[by default, regex that dedupes by matching the task ID]):
    submitted = _DataProcJob(other args..., job_dedupe_regex=job_dedupe_regex)
```

Comment (kaxil) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: @fenglu-g Can you take a look at this and let me know if you have any further comments?

Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: @fenglu-g Just wanted to clarify if my interpretation of your comment was in line with what you meant? If so I'll go ahead and implement it. Thanks!

Comment (fenglu-db) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: Sorry for the late reply, yes but we should also surface this job_dedupe_regex argument to dataproc submit job operators (e.g., [DataProcSparkOperator](https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/operators/dataproc_operator.py#L982)), wdyt?

**Comment by bolkedebruin** [comment]
Any updates guys?

**Comment by jj-ian** [comment]
Hi @bolkedebruin sorry for the delay; things got busy at work. Will try to make some headway on the requested changes this week.

File: airflow/contrib/hooks/gcp_dataproc_hook.py (hunk 1 above)
Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: Oh gotcha @fenglu-g, so users can provide their own regex (e.g. `DataProcSparkOperator(other args, [my_dedupe_regex_here]`) in `DataprocSparkOperator`, `DataprocPigOperator`, etc. I'll make that change. And sorry about my (very) late reply!

Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: Hi @fenglu-g , I started working on the regex idea and I don't think it makes sense to have the user define the deduping logic. 

It seems more error prone to me to have the user define the job ID they want to use for deduping, since the job ID is not user defined, but is instead hardcoded by Airflow here as the task ID + first 8 characters of UUID: https://github.com/apache/incubator-airflow/blob/0e8394fd23d067b7e226c011bb1825ff734219c5/airflow/contrib/hooks/gcp_dataproc_hook.py#L94. Since Airflow sets the job ID, Airflow should also set the deduping logic.

I understand your concern about not wanting to hardcode the deduping logic, though. A solution for this could be: instead of having Airflow hardcode the job ID as in the link above, to have it instead call a function that creates the job ID from the task ID. And then my deduping logic would reference an inverse of that function to extract the task ID for deduping instead of hardcoding the deduping logic. 

In other words, Airflow would create the job ID with a function that maps the task ID -> job ID. And then the dedupe mechanism would call a function that maps from job ID back to task ID to use in deduping. Tests will be added to make sure that these two functions are inverses of each other. 

How does that sound?

Comment (jj-ian) [reply airflow/contrib/hooks/gcp_dataproc_hook.py:52]: I'm going to go ahead with implementing the solution I proposed above @fenglu-g; if you have any reservations please let me know.

**Comment by feng-tao** [comment]
PTAL @fenglu-g  @bolkedebruin

**Comment by kurtqq** [comment]
This looks good to me.
But it needs rebase

**Comment by dossett** [comment]
@fenglu-g @bolkedebruin @feng-tao I can rebase this PR is there is interest in accepting it.  We (Etsy) have been running it for months and it works wonderfully, saving us time and money whenever Airflow restarts for whatever reason.

**Comment by potiuk** [comment]
@dossett  -> please rebase. I will review and take care about merging it.

**Comment by dossett** [comment]
@potiuk Thank you!  I pushed a rebase with an assist from @jj-ian

**Comment by potiuk** [comment]
Seems there are still some flake8 problems.

**Comment by dossett** [comment]
Thank @potiuk I think I got the last of them.

**Comment by dossett** [comment]
I don't understand the `Check docs` failure, it seems unrelated to this PR.

**Comment by mik-laj** [comment]
@dossett I restarted the job because it actually looks like some other server had existential problems. If the problem still exists, it means that everyone has this problem and you will have to do rebase when the patch appears on the master.

**Comment by mik-laj** [comment]
Job "Check docs" is green now. so we are waiting for the remaining jobs.

**Comment by mik-laj** [comment]
All jobs in Travis CI is green now.
